### ISSUE TITLE
Materialize Dictionaries in Group Keys

### ISSUE THREAD

**Issue by tustvold** [apache/arrow-datafusion#7647]
### Is your feature request related to a problem or challenge?

Currently grouping on a dictionary column will return dictionary-encoded group keys. Given that group keys inherently have few repeated values, especially when grouping on a single column, the use of dictionary encoding is unlikely to be yielding significant returns. Additionally following https://github.com/apache/arrow-datafusion/pull/7587 computing the dictionary is a non-trivial operation that could be eliminated

### Describe the solution you'd like

When grouping on a dictionary column, e.g. `Dictionary(DataType::Int32, DataType::Utf8)`, the returned schema should be the underlying value type, i.e. `DataType::Utf8`.

**Comment by qrilka** [comment]
Hi @tustvold 
I wanted to look into this ticket, I see the TODO you've left in #7587. Could you advise me some easy way to to simulate this situation e.g. in datafusion-cli or maybe there is some similar test in the code base?

**Comment by alamb** [comment]
You can great dictionary encoded columns using `arrow_cast`

Here is an example https://github.com/apache/arrow-datafusion/blob/e23d34bae60bb2f9c496241e218bab795af3af83/datafusion/sqllogictest/test_files/aggregate.slt#L2207-L2227

(though it might be helpful to use more than a single column)

**Comment by qrilka** [comment]
@tustvold I've spent some time getting familiar with the code for aggregations and I have a couple of questions. Maybe you could help me here?
I see that for aggregates in an initial plan we have in the base case 2 nested `AggregateExec`s: Partial and Final(Partitioned). And every `AggregateExec` uses `RowConverter` to convert arrays into rows for its input and back from rows into arrays for its output.
I assume that these conversions are basically inevitable because `RecordBatch`es are sent between execution stages and aggregation internally operates on rows, not arrays, is my understanding correct here?
It looks to me that it makes sense to convert from Dictionary to Utf8 on the first AggregateExec but it doesn't look like there's no mode like AggregateInitial. What could be the proper way out here? Maybe there could be a conversion step before aggregation added if any of the fields has a dictionary type? What would you advise?

**Comment by qrilka** [comment]
@alamb maybe you could help me here?

**Comment by tustvold** [comment]
You should not need to do any conversion, the conversion is already being done by RowConverter. Currently there is a cast annotated with a TODO linking back to this ticket. It should just be a case of removing this TODO and updating the schema logic within the plan to account for this

**Comment by qrilka** [comment]
@tustvold I saw the conversion in RowConverter but as I understand it happens multiple times: i.e. we convert dictionary into utf8 in AggregatePartial, then cast results back to dictionary and then we do the same in AggregateFinal.

Please let me know if I get your idea right: in the case of AggregatePartial followed by AggregateFinal, the schema for the first one should have Utf8 as its output and the second aggregate will then use Utf8 both as its input and as output.

**Comment by alamb** [comment]
TLDR I recommend we revert this change and reopen this ticket while we reconsider how to handle this case better.

# Background

This ticket caused a functional regression for us downstream in https://github.com/apache/arrow-datafusion/issues/8738 (a query that used to run started erroring). 

The cause of the issue is that the `LogicalPlan` and `ExecutionPlan`s schemas no longer match. I started exploring making them match in https://github.com/apache/arrow-datafusion/pull/8766

However while working on that issue,  I thought more about this change and I am not sure the change described in this issue is good for multi column groupings at all.

# Rationale

Dictionary Encoding on single column group keys is bad  because there is no repetition in the data and therefore the Dictionary encoding is pure over head. For example, given                                                                   
                                                                                     
```sql
SELECT ... GROUP BY a                                                                
```
Dictionary Encoding is worse than native encoding:

```
┌──────┐                   ┌──────┐      ┌────┐     
│ foo  │                   │ foo  │      │ 0  │     
├──────┤                   ├──────┤      ├────┤     
│ bar  │                   │ bar  │      │ 1  │     
├──────┤                   ├──────┤      ├────┤     
│ baz  │  ────────▶        │ baz  │      │ 2  │     
├──────┤                   ├──────┤      ├────┤     
│  ff  │                   │  ff  │      │ 3  │     
├──────┤                   ├──────┤      ├────┤     
│  ..  │                   │  ..  │      │ .. │     
├──────┤                   ├──────┤      ├────┤     
│ aaz  │                   │ aaz  │      │9999│     
└──────┘                   └──────┘      └────┘     
                                                    
 Group Values            values array   keys array  
   (distinct                                        
 values of a)                                       
```

However, the story is different when there is a multi column group key, as in that case, dictionary encoding each column can be a significant performance improvement as they are applied to each column individually and each column may have substantial redundancy. For example, given this query

```sql
SELECT ... GROUP BY a,b
```

```
┌──────┐  ┌──────┐               ┌──────┐      ┌────┐         ┌──────┐      ┌────┐      
│ foo  │  │ tag1 │               │ foo  │      │ 0  │         │ tag1 │      │ 0  │      
├──────┤  ├──────┤               ├──────┤      ├────┤         ├──────┤      ├────┤      
│ foo  │  │ tag2 │               │ baz  │      │ 0  │         │ tag2 │      │ 1  │      
├──────┤  ├──────┤               └──────┘      ├────┤         ├──────┤      ├────┤      
│ foo  │  │ tag3 │ ────────▶                   │ 0  │         │ tag3 │      │ 2  │      
├──────┤  ├──────┤                             ├────┤         ├──────┤      ├────┤      
│ foo  │  │ tag4 │                             │ 0  │         │ tag4 │      │ 3  │      
├──────┤  ├──────┤                             ├────┤         └──────┘      ├────┤      
│  ..  │  │  ..  │                             │ .. │                       │ .. │      
├──────┤  ├──────┤                             ├────┤                       ├────┤      
│ baz  │  │ tag4 │                             │ 1  │                       │ 3  │      
└──────┘  └──────┘                             └────┘                       └────┘      
                                                                                        
       Group Values            values array   keys array    values array   keys array   
    (distinct values of             (a)           (a)            (b)           (b)      
           a,b)                                                                         
```

This could especially impact multi-phase grouping where dictionary encoding will save significant time hashing values for low cardinality string columns.

In fact we think we may have seen a performance regression when picking up this change downstream as well, which could also be explained by the observation abive

Thus I recommend we revert this change via https://github.com/apache/arrow-datafusion/pull/8740 while we reconsider how to handle this case (maybe just for single column group by? Maybe do the dictionary encoding within the `RowEncoder` to avoid generating many redudant strings?

**Comment by tustvold** [comment]
> This could especially impact multi-phase grouping where dictionary encoding will save significant time hashing values for low cardinality string columns.

If anything this is precisely the case where the old behaviour was especially inefficient. It will compute the rows, expanding the dictionary, perform the grouping, convert the rows back to strings, convert the strings back to dictionaries, convert back to strings in the row format, again expanding the dictionaries, and repeat for each stage

**Comment by alamb** [comment]
I want to be clear that I have no particular evidence one way or the other about the performance implications of this particular change (and I probably confused the issue with speculation in https://github.com/apache/arrow-datafusion/issues/7647#issuecomment-1879161073)

So what I would like to do is:
1. Revert the change
2. Reopen this ticket
3. Work on getting a PR ready to put the change back in that both 1) doesn't cause a functional regression and 2) we have evidence improves performance

**Comment by qrilka** [comment]
@alamb what could be some evidence for 3.2? Is there anything in the code base or maybe in some other ticket? The plan you show makes total sense

**Comment by alamb** [comment]
Thanks @qrilka 

> @alamb what could be some evidence for 3.2? 

I think we need to create a benchmark that does aggregation queries on dictionary encoded string columns (I know the existing end to end TPCH, and ClickBench benchmarks do not do this). I will file a ticket shortly about this

**Comment by alamb** [comment]
I filed https://github.com/apache/arrow-datafusion/issues/8791 to track adding additional test coverage, and I plan to work on that in the next few days

**Comment by ghostiee-11** [comment]
Related to this issue: on datafusion 54 a dictionary group key whose combined cardinality across partitions exceeds its key width fails hard, rather than just being suboptimal.

## Repro (pure datafusion + pyarrow, no network)

```python
import numpy as np, pyarrow as pa
from datafusion import SessionContext

PER_PARTITION, N_PARTITIONS = 100, 50  # combined 5000 >> Int8 max (127)
parts = []
for p in range(N_PARTITIONS):
    values = pa.array(np.arange(p*PER_PARTITION, p*PER_PARTITION+PER_PARTITION, dtype=np.int64))
    keys = pa.array(np.arange(PER_PARTITION, dtype=np.int8))
    k = pa.DictionaryArray.from_arrays(keys, values)  # valid Dictionary(Int8, Int64) per batch
    parts.append(pa.record_batch({"k": k, "v": pa.array(np.ones(PER_PARTITION))}))

ctx = SessionContext()
ctx.register_record_batches("t", [[b] for b in parts])
ctx.sql("SELECT k, SUM(v) FROM t GROUP BY k").to_arrow_table()
# -> Arrow error: Dictionary key bigger than the key type
```

Each partition is a valid `Dictionary(Int8, Int64)` (<=128 distinct), but the partitions hold disjoint values, so the combined group-key set is 5000.

## Mechanism (datafusion 54)

- `Aggregate::group_fields` (physical-plan/src/aggregates/mod.rs) keeps the dictionary type for the group key.
- `GroupValuesRows::emit` decodes the keys via the RowConverter, then `dictionary_encode_if_necessary` (group_values/row.rs) re-encodes them back into the narrow dictionary because the group schema says so.
- Across `RepartitionExec: Hash` / `FinalPartitioned`, the combined set exceeds Int8, so the re-encode overflows. arrow-rs is behaving correctly here (`concat` is type-preserving, so it errors in `merge_dictionary_values` rather than silently widening); the decision to keep a narrow key for the combined group keys is the datafusion side.

`#8291` materialized dict group keys to the value type and avoided this; that path is not present in 54. This issue and `#21765` are moving toward keeping dictionaries and optimizing on them, which is great for low cardinality but needs to handle the case where the combined cardinality outgrows the key width.

## Question before I send a PR

Two ways to fix it, and I would rather match your direction:

- **A. Decode** dict group keys to their value type at plan time (restores `#8291` semantics). Simple, output type changes to the value type.
- **B. Widen** the group key to a safe key width (e.g. Int32/Int64) and keep a dictionary output (aligns with `#21765`). Preserves "group by dict returns dict".

I have both a candidate patch and a sqllogictest ready. Which direction do you prefer for the aggregate group-key path? Happy to open the PR against whichever you pick.

**Comment by Rich-T-kid** [comment]
> Related to this issue: on datafusion 54 a dictionary group key whose combined cardinality across partitions exceeds its key width fails hard, rather than just being suboptimal.
> 
> ## Repro (pure datafusion + pyarrow, no network)
> import numpy as np, pyarrow as pa
> from datafusion import SessionContext
> 
> PER_PARTITION, N_PARTITIONS = 100, 50  # combined 5000 >> Int8 max (127)
> parts = []
> for p in range(N_PARTITIONS):
>     values = pa.array(np.arange(p*PER_PARTITION, p*PER_PARTITION+PER_PARTITION, dtype=np.int64))
>     keys = pa.array(np.arange(PER_PARTITION, dtype=np.int8))
>     k = pa.DictionaryArray.from_arrays(keys, values)  # valid Dictionary(Int8, Int64) per batch
>     parts.append(pa.record_batch({"k": k, "v": pa.array(np.ones(PER_PARTITION))}))
> 
> ctx = SessionContext()
> ctx.register_record_batches("t", [[b] for b in parts])
> ctx.sql("SELECT k, SUM(v) FROM t GROUP BY k").to_arrow_table()
> # -> Arrow error: Dictionary key bigger than the key type
> Each partition is a valid `Dictionary(Int8, Int64)` (<=128 distinct), but the partitions hold disjoint values, so the combined group-key set is 5000.
> 
> ## Mechanism (datafusion 54)
> * `Aggregate::group_fields` (physical-plan/src/aggregates/mod.rs) keeps the dictionary type for the group key.
> * `GroupValuesRows::emit` decodes the keys via the RowConverter, then `dictionary_encode_if_necessary` (group_values/row.rs) re-encodes them back into the narrow dictionary because the group schema says so.
> * Across `RepartitionExec: Hash` / `FinalPartitioned`, the combined set exceeds Int8, so the re-encode overflows. arrow-rs is behaving correctly here (`concat` is type-preserving, so it errors in `merge_dictionary_values` rather than silently widening); the decision to keep a narrow key for the combined group keys is the datafusion side.
> 
> `#8291` materialized dict group keys to the value type and avoided this; that path is not present in 54. This issue and `#21765` are moving toward keeping dictionaries and optimizing on them, which is great for low cardinality but needs to handle the case where the combined cardinality outgrows the key width.
> 
> ## Question before I send a PR
> Two ways to fix it, and I would rather match your direction:
> 
> * **A. Decode** dict group keys to their value type at plan time (restores `#8291` semantics). Simple, output type changes to the value type.
> * **B. Widen** the group key to a safe key width (e.g. Int32/Int64) and keep a dictionary output (aligns with `#21765`). Preserves "group by dict returns dict".
> 
> I have both a candidate patch and a sqllogictest ready. Which direction do you prefer for the aggregate group-key path? Happy to open the PR against whichever you pick.

you may be interested in this issue https://github.com/apache/datafusion/issues/23127

**Comment by Rich-T-kid** [comment]
@alamb I believe https://github.com/apache/datafusion/pull/23187 closed this issue since it removes the cast round trip for dictionary arrays in `GroupValues`
